File: cpp/236.cpp
#include<vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
  void dfs(TreeNode* root, TreeNode* p, TreeNode* q,
           vector<TreeNode*> &cur, vector<vector<TreeNode*>> &res){
    cur.push_back(root);
    if(root == p || root == q)
      res.push_back(vector<TreeNode*>(cur.begin(), cur.end()));
    if(root->left) dfs(root->left, p, q, cur, res);
    if(root->right) dfs(root->right, p, q, cur, res);
    cur.pop_back();
  }
public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    vector<TreeNode*> cur;
    vector<vector<TreeNode*>> res;
    dfs(root, p, q, cur, res);
    TreeNode* ans;
    int i = 0;
    while(res[0][i] == res[1][i]){
      ans = res[0][i]; ++i;
    }
    return ans;
  }
};

File: python/236.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def dfs(self, root, cur, p, q, res):
        cur.append(root)
        if root == p or root == q:
            res.append(cur[:])
        if root.left:
            self.dfs(root.left, cur, p, q, res)
        if root.right:
            self.dfs(root.right, cur, p, q, res)
        cur.pop()

    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        res = []
        self.dfs(root, [], p, q, res)
        return [i for i, j in zip(res[0], res[1]) if i == j][-1]
